### PR TITLE
Don't split reductions to remove nested thread mapping

### PR DESCRIPTION
**Pull request by ftynse**
Reduction detection mechanism splits out the band containing the reduction loop into a separate band in the tree.  As a result, mapping to threads was made recursive to allow for mapping both the reduction update band and its parent band node, although mapping nested bands to threads is invalid in the general case.  However, it is necessary to split out the reduction loop in order to insert the reduction synchronization call.

Split out the reduction loop immediately before inserting the reduction synchronization.  During thread mapping, keep the reduction loop inside its original band.  This avoids the need to keep track of the relation between the reduction band and its parent, which can be replaced by position of the reduction loop in the band.  As memory promotion may split the band to insert copies, insert reduction synchronizations before promotion passes.

mapInnermostBandsToThreads maps coincident band members to threads while traversing the tree recursively in DFS postorder.  It is possible that members of two nested bands get mapped to thread dimensions, which was necessary for parallel reduction support.  The previous commit keeps the reduction in its original band, removing the need for mapping nested bands. Such mapping is not valid in general because members of different bands are not necessarily permutable with each other.  Since thread dimensions are not executed in any specific order, i.e. there is no guarantee that, for a given value of threadIdx.y, all threadIdx.x threads will be executed before (or after) those with another value of threadIdx.y, permutability is a necessary condition for mapping to multiple thread dimensions.  Only allow mapping a band to threads if its children are not already mapped to threads.  The procedure continues recursively so as to insert all relevant synchronization between children of sequence nodes even after the thread mapping is complete.

Closes #263

### CODE REVIEW

File: src/core/polyhedral/cuda/mapped_scop.cc [hunk 1]
```diff
@@ -220,65 +220,52 @@ bool MappedScop::detectReductions(detail::ScheduleTree* tree) {
...
   // Order the init statements (if any) before the update statements
   // to ensure the band from which the reduction band has been split off
   // only contains update statements.
   // Note that this relies on the outer members being coincident.
   if (!inits.is_empty()) {
     orderBefore(scop_->scheduleRoot(), tree, inits);
   }
-  reductionFromParent_.emplace(tree, reductionTree);
-  reductionBandUpdates_.emplace(reductionTree, updates);
+  reductionBandUpdates_.emplace(tree, Reduction(updates, reductionDim));
```
Comment (skimo-openhub) [inline src/core/polyhedral/cuda/mapped_scop.cc]: Update documentation of `detectReductions` corresponding to these changes.

File: src/core/polyhedral/cuda/mapped_scop.cc [hunk 2]
```diff
@@ -220,65 +220,52 @@ bool MappedScop::detectReductions(detail::ScheduleTree* tree) {
...
   CHECK(reductionBandUpdates_.count(st) == 1);
-
   auto reductionBand = st->elemAs<detail::ScheduleTreeElemBand>();
   CHECK(reductionBand);
-  // Start with the schedule of the reduction band (in last position).
-  auto reductionSchedule = reductionBand->mupa_;
 
   // Total size of returned schedule needs to be equal
   // to the number of thread identifiers.
-  if (numThreads.view.size() > 1) {
```
Comment (skimo-openhub) [inline src/core/polyhedral/cuda/mapped_scop.cc]: The comment above is about the removed code, so it should presumably be removed too.

File: src/core/polyhedral/cuda/mapped_scop.cc [hunk 3]
```diff
@@ -329,10 +316,8 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
 size_t MappedScop::mapToThreads(detail::ScheduleTree* band, size_t nInner) {
   using namespace tc::polyhedral::detail;
 
-  if (nInner >= numThreads.view.size()) {
-    return nInner;
-  }
```
Comment (skimo-openhub) [inline src/core/polyhedral/cuda/mapped_scop.cc]: Why did you remove this test?
In fact, it should be strengthened to `nInner > 0`, no?

File: src/core/polyhedral/cuda/mapped_scop.cc [hunk 4]
```diff
@@ -329,10 +316,8 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
...
-  if (reductionBandUpdates_.count(band) == 1) {
+  bool mapReduction = reductionBandUpdates_.count(band) == 1;
+  if (mapReduction) {
     // A reduction is assumed to get mapped to threadIdx.x
     if (nInner != 0) {
```
Comment (skimo-openhub) [inline src/core/polyhedral/cuda/mapped_scop.cc]: This test can then be removed.

**Comment by skimo-openhub** [comment]
On Tue, Apr 10, 2018 at 01:53:28AM -0700, skimo-openhub wrote:
> skimo-openhub commented on this pull request.

I keep forgetting that github doesn't seem to keep track
of the commit that a review corresponds to.
This was cdb9c07d6af6542efd3ecfcd6c3977f836d5da4f.

skimo

File: src/core/polyhedral/cuda/mapped_scop.cc (hunk 3 above)
Comment (ftynse) [reply src/core/polyhedral/cuda/mapped_scop.cc]: `nInner` is completely removed in the following commit, I may have picked the wrong change

File: src/core/polyhedral/cuda/mapped_scop.cc (hunk 4 above)
Comment (ftynse) [reply src/core/polyhedral/cuda/mapped_scop.cc]: It will be removed in the next commit.  The first one only deals with separating the reduction band without changing how the threads are mapped in general.

**Comment by ftynse** [comment]
updated

**Comment by ftynse** [comment]
rebased

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 5]
```diff
@@ -225,65 +225,50 @@ bool MappedScop::detectReductions(detail::ScheduleTree* tree) {
...
+  auto nMember = reductionBand->nMember();
+  auto reductionDim = reductionBandUpdates_.at(st).reductionDim;
+  auto nMappedThreads =
+      std::min(numThreads.view.size(), reductionBand->nOuterCoincident() + 1);
+  CHECK_GE(nMember, reductionDim);
+  CHECK_GE(reductionDim + 1, nMappedThreads);
+  reductionSchedule = reductionSchedule.drop_dims(
+      isl::dim_type::set, reductionDim + 1, nMember - reductionDim - 1);
+  reductionSchedule = reductionSchedule.drop_dims(
+      isl::dim_type::set, 0, reductionDim - nMappedThreads + 1);
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: nit: I would put the `+ 1` together with `reductionDim` and
probably also write `-(reductionDim + 1)` instead of `- reductionDim - 1`.
Note that I will probably end up rewriting these lines anyway in the not too distant future (because of the `isl::dim_type::set`).

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 6]
```diff
@@ -587,6 +575,22 @@ std::tuple<std::string, tc::Grid, tc::Block> MappedScop::codegen(
       mappedScopForCodegen->numThreads);
 }
 
+// Split out reduction loops into separate bands and insert reduction
+// synchronizations after those bands.
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: "after" is confusing.  It's inserted _outside_ of those bands, isn't it?

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 7]
```diff
@@ -587,6 +575,22 @@ std::tuple<std::string, tc::Grid, tc::Block> MappedScop::codegen(
...
+void MappedScop::insertReductionSyncs() {
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: Maybe rename the method to reflect the two operation?
`splitOutReductionsAndInsertSyncs`?

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 8]
```diff
@@ -316,62 +312,49 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
...
-        band->child({0}), nInner, numThreads.view.size());
-    scop_->insertSyncAfter(band->child({0}));
-    return numThreads.view.size();
-  }
   // With current isl scheduler, if coincident dimensions exist in a band,
   // they are outermost.
   // If a band has more than 3 coincident dimensions, this will choose
   // outermost, but we may also want innermost.
   auto nOuterCoincident = bandNode->nOuterCoincident();
+  int firstThreadIndex = mapReduction ? 1 : 0;
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: I find this name (`firstThreadIndex`) confusing.
It seems to indicate the _number_ of threads already used.
I'd either put it in `nMappedThreads` or in a new `nMappedReductionThreads`.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 9]
```diff
@@ -316,62 +312,49 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
...
   if (!bandNode->permutable_ || nOuterCoincident < 1) {
-    return nInner;
+    return firstThreadIndex;
   }
 
   auto nMappedThreads = std::min(
-      numThreads.view.size() - nInner, static_cast<size_t>(nOuterCoincident));
-  CHECK_GT(nMappedThreads, 0) << "not mapping to threads";
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: You removed this check.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 10]
```diff
@@ -316,62 +312,49 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
...
-    CHECK_EQ(band->numChildren(), 1);
-    mapRemaining<mapping::ThreadId>(
-        band->child({0}), nInner, numThreads.view.size());
-    scop_->insertSyncAfter(band->child({0}));
-    return numThreads.view.size();
-  }
   // With current isl scheduler, if coincident dimensions exist in a band,
   // they are outermost.
   // If a band has more than 3 coincident dimensions, this will choose
   // outermost, but we may also want innermost.
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: This comment looks outdated (prior to this series).
This series would be a good time to fix it, because the reasoning is hard to follow now.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 11]
```diff
@@ -316,62 +312,49 @@ detail::ScheduleTree* MappedScop::separateReduction(detail::ScheduleTree* st) {
...
   auto nOuterCoincident = bandNode->nOuterCoincident();
+  int firstThreadIndex = mapReduction ? 1 : 0;
   if (!bandNode->permutable_ || nOuterCoincident < 1) {
-    return nInner;
+    return firstThreadIndex;
```
Comment (skimo-openhub) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: I think this deserves a bit more explanation.
It is only correct because `MappedScop::detectReductions` has already done
the required checks.  Actually, it hasn't completely.
`MappedScop::detectReductions` doesn't have the permutable_` check.
That should be fixed first.

File: tc/core/polyhedral/cuda/mapped_scop.cc (hunk 9 above)
Comment (ftynse) [reply tc/core/polyhedral/cuda/mapped_scop.cc]: I replaced it by early return, because `nMappedThreads` can now legitimately be zero in case when reduction was mapped.

**Comment by ftynse** [comment]
> Now that we only map a single band, we might as well complete the mapping immediately instead of waiting until we reach the parent, but this can be done in a follow-up.

I also thought about that.  Updated.

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 12]
```diff
@@ -467,20 +467,22 @@ size_t MappedScop::mapInnermostBandsToThreads(detail::ScheduleTree* st) {
   if (auto band = st->elemAs<detail::ScheduleTreeElemBand>()) {
     if (n == 0) {
       // If children were not mapped to threads, the current band can be mapped.
+      // First, map the coincidence and reduction dimension to threads.
+      // The fix unused thread dimensions to 0 because we cannot map parent
+      // bands anyway.
       auto nMapped = mapToThreads(st);
+      mapRemaining<mapping::ThreadId>(st, nMapped, numThreads.view.size());
```
[the remark this answers is not in the document]
Comment (ftynse) [reply tc/core/polyhedral/cuda/mapped_scop.cc]: Good point about "if anything gets mapped", but I'm not convinced it should be inside `mapToThreads`.  The function has two exit points and we'd have to eventually map to remaining at both.

**Comment by ftynse** [comment]
Thanks, running tests...

**Comment by ftynse** [comment]
tests pass, merging
